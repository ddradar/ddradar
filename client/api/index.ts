export const apiPrefix = '/api/v1'
