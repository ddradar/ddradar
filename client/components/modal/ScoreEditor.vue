<template>
  <div class="modal-card">
    <header class="modal-card-head">
      <h1 class="modal-card-title">{{ songData.name }}</h1>
    </header>
    <section class="modal-card-body">
      <!-- Select chart -->
      <h2 v-if="selectedChart" class="subtitle is-small">
        {{ chartName }}
      </h2>
      <b-field v-else :label="$t('label.selectChart')">
        <b-select :disabled="selectedChart" @input="onChartSelected">
          <option v-for="c in charts" :key="c.label" :value="c">
            {{ c.label }}
          </option>
        </b-select>
      </b-field>

      <!-- Input score -->
      <template v-if="selectedChart">
        <b-loading :active.sync="isLoading" />
        <b-field grouped>
          <b-field :label="$t('label.score')">
            <b-input
              v-model.number="score"
              type="number"
              required
              placeholder="0-1000000"
              min="0"
              max="1000000"
              step="10"
            />
          </b-field>
          <b-checkbox v-model="isFailed">{{ $t('label.rankE') }}</b-checkbox>
        </b-field>

        <b-field :label="$t('label.clear')">
          <b-select v-model.number="clearLamp" :placeholder="$t('label.clear')">
            <option value="0">Failed</option>
            <option value="1">Assisted Clear</option>
            <option value="2">Clear</option>
            <option value="3">Life4 Clear</option>
            <option value="4">Good Full Combo</option>
            <option value="5">Great Full Combo</option>
            <option value="6">Perfect Full Combo</option>
            <option value="7">Marvelous Full Combo</option>
          </b-select>
        </b-field>

        <b-field :label="$t('label.exScore')">
          <b-input
            v-model.number="exScore"
            type="number"
            min="0"
            :max="exScoreMax"
            :placeholder="`0-${exScoreMax}`"
          />
        </b-field>

        <b-field :label="$t('label.maxCombo')">
          <b-input
            v-model.number="maxCombo"
            type="number"
            min="0"
            :max="maxComboMax"
            :placeholder="`0-${maxComboMax}`"
          />
        </b-field>

        <b-field>
          <b-button type="is-info" icon-left="calculator" @click="calcScore()">
            {{ $t('button.calc') }}
          </b-button>
        </b-field>
      </template>
    </section>

    <footer v-if="selectedChart" class="modal-card-foot">
      <b-button type="is-success" icon-left="content-save" @click="saveScore()">
        {{ $t('button.save') }}
      </b-button>
      <b-button type="is-danger" icon-left="delete" @click="deleteScore()">
        {{ $t('button.delete') }}
      </b-button>
      <b-button type="is-warning" @click="close()">
        {{ $t('button.close') }}
      </b-button>
    </footer>
  </div>
</template>

<i18n>
{
  "ja": {
    "label": {
      "selectChart": "譜面を選択",
      "score": "ハイスコア",
      "rankE": "E判定",
      "clear": "クリア種別",
      "exScore": "EXスコア",
      "maxCombo": "最大コンボ数"
    },
    "button": {
      "calc": "自動計算",
      "save": "保存",
      "delete": "削除",
      "close": "閉じる"
    },
    "message": {
      "cannotCalc": "情報が足りないため、スコアの自動計算ができませんでした。",
      "successSave": "保存しました",
      "deleteScore": "スコアを削除しますか？",
      "successDelete": "削除しました"
    }
  },
  "en": {
    "label": {
      "selectChart": "Select Chart",
      "score": "Score",
      "rankE": "Rank E",
      "clear": "Clear Lamp",
      "exScore": "EX SCORE",
      "maxCombo": "MAX COMBO"
    },
    "button": {
      "calc": "Auto Calc",
      "save": "Save",
      "delete": "Delete",
      "close": "Close"
    },
    "message": {
      "cannotCalc": "Cannot guess score due to lack of information.",
      "successSave": "Saved",
      "deleteScore": "Do you delete this score?",
      "successDelete": "Deleted"
    }
  }
}
</i18n>

<script lang="ts">
import type { Database } from '@ddradar/core'
import { Score, Song } from '@ddradar/core'
import { Component, Prop, Vue } from 'nuxt-property-decorator'

import { deleteChartScore, getChartScore, postChartScore } from '~/api/score'
import * as popup from '~/utils/popup'

type ChartSchema = Omit<
  Database.StepChartSchema | Database.CourseChartSchema,
  'level'
>
type SongOrCourseInfo = { name: string; charts: ChartSchema[] }

@Component({ fetchOnServer: false })
export default class ScoreEditorComponent extends Vue {
  @Prop({ required: true, type: String })
  readonly songId!: string

  @Prop({ required: false, type: Number, default: null })
  readonly playStyle!: Song.PlayStyle | null

  @Prop({ required: false, type: Number, default: null })
  readonly difficulty!: Song.Difficulty | null

  @Prop({ required: true, type: Object })
  readonly songData!: SongOrCourseInfo

  score = 0
  exScore = 0
  clearLamp: Score.ClearLamp = 0
  maxCombo = 0
  isFailed = false

  selectedChart: ChartSchema | null = null

  isLoading = true

  get rank() {
    return this.isFailed ? 'E' : Score.getDanceLevel(this.score)
  }

  get exScoreMax() {
    if (!this.selectedChart) return 0
    return (
      (this.selectedChart.notes +
        this.selectedChart.freezeArrow +
        this.selectedChart.shockArrow) *
      3
    )
  }

  get chartName() {
    if (!this.selectedChart) return null
    const playStyle = Song.playStyleMap.get(this.selectedChart.playStyle)
    const difficulty = Song.difficultyMap.get(this.selectedChart.difficulty)
    return `${playStyle}/${difficulty}`
  }

  get maxComboMax() {
    if (!this.selectedChart) return 0
    return this.selectedChart.notes + this.selectedChart.shockArrow
  }

  get charts() {
    return this.songData.charts.map(c => ({
      playStyle: c.playStyle,
      difficulty: c.difficulty,
      label: `${Song.playStyleMap.get(c.playStyle)}/${Song.difficultyMap.get(
        c.difficulty
      )}`,
    }))
  }

  async created() {
    if (this.playStyle !== null && this.difficulty !== null) {
      this.selectedChart =
        this.songData.charts.find(
          c =>
            c.playStyle === this.playStyle && c.difficulty === this.difficulty
        ) ?? null
      await this.fetchScore()
    }
  }

  async onChartSelected({
    playStyle,
    difficulty,
  }: Pick<ChartSchema, 'playStyle' | 'difficulty'>) {
    this.selectedChart =
      this.songData.charts.find(
        c => c.playStyle === playStyle && c.difficulty === difficulty
      ) ?? null
    await this.fetchScore()
  }

  calcScore() {
    /* istanbul ignore if */
    if (!this.selectedChart) return
    try {
      const score = Score.setValidScoreFromChart(this.selectedChart, {
        score: this.score,
        exScore: this.exScore,
        maxCombo: this.maxCombo,
        clearLamp: this.clearLamp,
        rank: this.isFailed ? 'E' : undefined,
      })
      this.score = score.score
      this.exScore = score.exScore ?? this.exScore
      this.maxCombo = score.maxCombo ?? this.maxCombo
      this.clearLamp = score.clearLamp
      this.isFailed = score.rank === 'E'
    } catch {
      popup.warning(this.$buefy, this.$t('message.cannotCalc') as string)
    }
  }

  async saveScore() {
    /* istanbul ignore if */
    if (!this.selectedChart) return
    const playStyle = this.selectedChart.playStyle
    const difficulty = this.selectedChart.difficulty
    try {
      await postChartScore(this.$http, this.songId, playStyle, difficulty, {
        score: this.score,
        exScore: this.exScore,
        maxCombo: this.maxCombo,
        clearLamp: this.clearLamp,
        rank: this.rank,
      })
      popup.success(this.$buefy, this.$t('message.successSave') as string)
    } catch (error) {
      popup.danger(this.$buefy, error.message ?? error)
    }
    this.close()
  }

  deleteScore() {
    this.$buefy.dialog.confirm({
      message: this.$t('message.deleteScore') as string,
      type: 'is-warning',
      hasIcon: true,
      onConfirm: async () => {
        await this.callDeleteAPI()
        this.close()
      },
    })
  }

  async fetchScore() {
    if (!this.selectedChart) return
    this.isLoading = true
    const playStyle = this.selectedChart.playStyle
    const difficulty = this.selectedChart.difficulty
    try {
      const scores = await getChartScore(
        this.$http,
        this.songId,
        playStyle,
        difficulty,
        'private'
      )
      this.score = scores[0].score
      this.exScore = scores[0].exScore ?? this.exScore
      this.clearLamp = scores[0].clearLamp
      this.maxCombo = scores[0].maxCombo ?? this.maxCombo
      this.isFailed = scores[0].rank === 'E'
    } catch (error) {
      this.isLoading = false
      const message = error.message ?? error
      if (message !== '404') {
        popup.danger(this.$buefy, message)
      }
    }
    this.isLoading = false
  }

  private async callDeleteAPI() {
    /* istanbul ignore if */
    if (!this.selectedChart) return
    const playStyle = this.selectedChart.playStyle
    const difficulty = this.selectedChart.difficulty
    try {
      await deleteChartScore(this.$http, this.songId, playStyle, difficulty)
      popup.success(this.$buefy, this.$t('message.successDelete') as string)
    } catch (error) {
      popup.danger(this.$buefy, error.message ?? error)
    }
  }

  close() {
    // @ts-ignore
    this.$parent.close()
  }
}
</script>
