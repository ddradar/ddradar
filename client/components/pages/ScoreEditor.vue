<template>
  <div class="modal-card">
    <header class="modal-card-head">
      <h1 class="modal-card-title">{{ songData.name }}</h1>
    </header>
    <section class="modal-card-body">
      <!-- Select chart -->
      <h2 v-if="selectedChart" class="subtitle is-small">
        {{ chartName }}
      </h2>
      <b-field v-else label="Select chart">
        <b-select :disabled="selectedChart" @input="onChartSelected">
          <option v-for="c in charts" :key="c.label" :value="c">
            {{ c.label }}
          </option>
        </b-select>
      </b-field>

      <!-- Input score -->
      <template v-if="selectedChart">
        <b-loading :active.sync="isLoading" />
        <b-field grouped>
          <b-field label="Score">
            <b-input
              v-model.number="score"
              type="number"
              required
              placeholder="0-1000000"
              min="0"
              max="1000000"
              step="10"
            />
          </b-field>
          <b-checkbox v-model="isFailed">E判定</b-checkbox>
        </b-field>

        <b-field label="Clear Lamp">
          <b-select v-model.number="clearLamp" placeholder="Clear Lamp">
            <option value="0">Failed</option>
            <option value="1">Assisted Clear</option>
            <option value="2">Clear</option>
            <option value="3">Life4 Clear</option>
            <option value="4">Good Full Combo</option>
            <option value="5">Great Full Combo</option>
            <option value="6">Perfect Full Combo</option>
            <option value="7">Marvelous Full Combo</option>
          </b-select>
        </b-field>

        <b-field label="EX SCORE">
          <b-input
            v-model.number="exScore"
            type="number"
            min="0"
            :max="exScoreMax"
            :placeholder="`0-${exScoreMax}`"
          />
        </b-field>

        <b-field label="MAX COMBO">
          <b-input
            v-model.number="maxCombo"
            type="number"
            min="0"
            :max="maxComboMax"
            :placeholder="`0-${maxComboMax}`"
          />
        </b-field>

        <b-field>
          <b-button type="is-info" icon-left="calculator" @click="calcScore()">
            自動計算
          </b-button>
        </b-field>
      </template>
    </section>

    <footer
      v-if="playStyle !== null && difficulty !== null"
      class="modal-card-foot"
    >
      <b-button type="is-success" icon-left="content-save" @click="saveScore()">
        Save
      </b-button>
      <b-button type="is-danger" icon-left="delete" @click="deleteScore()">
        Delete
      </b-button>
    </footer>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'nuxt-property-decorator'

import {
  ClearLamp,
  getDanceLevel,
  setValidScoreFromChart,
  UserScore,
} from '~/types/api/score'
import {
  getDifficultyName,
  getPlayStyleName,
  SongInfo,
  StepChart,
} from '~/types/api/song'

@Component({ fetchOnServer: false })
export default class ScoreEditorComponent extends Vue {
  @Prop({ required: true, type: String })
  readonly songId: string

  @Prop({ required: false, type: Number, default: null })
  readonly playStyle: 1 | 2 | null

  @Prop({ required: false, type: Number, default: null })
  readonly difficulty: 0 | 1 | 2 | 3 | 4 | null

  @Prop({ required: true, type: Object })
  readonly songData: SongInfo

  score = 0
  exScore = 0
  clearLamp: ClearLamp = 0
  maxCombo = 0
  isFailed = false

  selectedChart: StepChart | null = null

  isLoading = true

  get rank() {
    return this.isFailed ? 'E' : getDanceLevel(this.score)
  }

  get exScoreMax() {
    if (!this.selectedChart) return 0
    return (
      (this.selectedChart.notes +
        this.selectedChart.freezeArrow +
        this.selectedChart.shockArrow) *
      3
    )
  }

  get chartName() {
    if (!this.selectedChart) return null
    const playStyle = getPlayStyleName(this.selectedChart.playStyle)
    const difficulty = getDifficultyName(this.selectedChart.difficulty)
    return `${playStyle}/${difficulty}`
  }

  get maxComboMax() {
    if (!this.selectedChart) return 0
    return this.selectedChart.notes + this.selectedChart.shockArrow
  }

  get charts() {
    return this.songData.charts.map(c => ({
      playStyle: c.playStyle,
      difficulty: c.difficulty,
      label: `${getPlayStyleName(c.playStyle)}/${getDifficultyName(
        c.difficulty
      )}`,
    }))
  }

  async created() {
    if (this.playStyle !== null && this.difficulty !== null) {
      this.selectedChart =
        this.songData.charts.find(
          c =>
            c.playStyle === this.playStyle && c.difficulty === this.difficulty
        ) ?? null
      await this.fetchScore()
    }
  }

  async onChartSelected({
    playStyle,
    difficulty,
  }: Pick<StepChart, 'playStyle' | 'difficulty'>) {
    this.selectedChart = this.songData.charts.find(
      c => c.playStyle === playStyle && c.difficulty === difficulty
    )
    await this.fetchScore()
  }

  calcScore() {
    try {
      const score = setValidScoreFromChart(this.selectedChart, {
        score: this.score,
        exScore: this.exScore,
        maxCombo: this.maxCombo,
        clearLamp: this.clearLamp,
        rank: this.isFailed ? 'E' : undefined,
      })
      this.score = score.score
      this.exScore = score.exScore
      this.maxCombo = score.maxCombo
      this.clearLamp = score.clearLamp
      this.isFailed = score.rank === 'E'
    } catch {
      this.$buefy.notification.open({
        message: '情報が足りないため、スコアの自動計算ができませんでした。',
        type: 'is-warning',
        position: 'is-top',
        hasIcon: true,
      })
    }
  }

  async saveScore() {
    if (!this.selectedChart) return
    const playStyle = this.selectedChart.playStyle
    const difficulty = this.selectedChart.difficulty
    try {
      await this.$http.$post(
        `/api/v1/scores/${this.songId}/${playStyle}/${difficulty}`,
        {
          score: this.score,
          exScore: this.exScore,
          maxCombo: this.maxCombo,
          clearLamp: this.clearLamp,
          rank: this.rank,
        }
      )
      this.$buefy.notification.open({
        message: 'Success!',
        type: 'is-success',
        position: 'is-top',
        hasIcon: true,
      })
    } catch (error) {
      this.$buefy.notification.open({
        message: error.message ?? error,
        type: 'is-danger',
        position: 'is-top',
        hasIcon: true,
      })
    }
    // @ts-ignore
    this.$parent.close()
  }

  deleteScore() {
    this.$buefy.dialog.confirm({
      message: 'スコアを削除しますか？',
      type: 'is-warning',
      hasIcon: true,
      onConfirm: async () => {
        await this.callDeleteAPI()
        // @ts-ignore
        this.$parent.close()
      },
    })
  }

  async fetchScore() {
    if (!this.selectedChart) return
    this.isLoading = true
    const playStyle = this.selectedChart.playStyle
    const difficulty = this.selectedChart.difficulty
    try {
      const scores = await this.$http.$get<UserScore[]>(
        `/api/v1/scores/${this.songId}/${playStyle}/${difficulty}?scope=private`
      )
      this.score = scores[0].score
      this.exScore = scores[0].exScore
      this.clearLamp = scores[0].clearLamp
      this.maxCombo = scores[0].maxCombo
      this.isFailed = scores[0].rank === 'E'
    } catch (error) {
      this.isLoading = false
      const message = error.message ?? error
      if (message !== '404') {
        this.$buefy.notification.open({
          message,
          type: 'is-danger',
          position: 'is-top',
          hasIcon: true,
        })
      }
    }
    this.isLoading = false
  }

  async callDeleteAPI() {
    if (!this.selectedChart) return
    const playStyle = this.selectedChart.playStyle
    const difficulty = this.selectedChart.difficulty
    try {
      await this.$http.delete(
        `/api/v1/scores/${this.songId}/${playStyle}/${difficulty}`
      )
      this.$buefy.notification.open({
        message: 'Success!',
        type: 'is-success',
        position: 'is-top',
        hasIcon: true,
      })
    } catch (error) {
      this.$buefy.notification.open({
        message: error.message ?? error,
        type: 'is-danger',
        position: 'is-top',
        hasIcon: true,
      })
    }
  }
}
</script>
