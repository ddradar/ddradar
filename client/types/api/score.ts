/**
 * Object type returned by `/api/v1/scores/{:songId}/{:playStyle}/{:difficulty}`
 * @see https://github.com/ddradar/ddradar/blob/master/api/getChartScore/README.md
 */
export type UserScore = {
  userId: string
  userName: string
  /** Song id that depend on official site. ^([01689bdiloqDIOPQ]*){32}$ */
  songId: string
  songName: string
  playStyle: 1 | 2
  difficulty: 0 | 1 | 2 | 3 | 4
  /** Normal score */
  score: number
  exScore?: number
  maxCombo?: number
  /** 0: Failed, 1: Assisted Clear 2: Clear, 3: LIFE4, 4: Good FC (Full Combo), 5: Great FC, 6: PFC, 7: MFC */
  clearLamp: ClearLamp
  /** Clear rank (E-AAA) */
  rank: string
}

/** 0: Failed, 1: Assisted Clear 2: Clear, 3: LIFE4, 4: Good FC (Full Combo), 5: Great FC, 6: PFC, 7: MFC */
export type ClearLamp = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7

/**
 * Request body to `/api/v1/scores/{:songId}/{:playStyle}/{:difficulty}`
 * @see https://github.com/ddradar/ddradar/blob/master/api/postChartScore/README.md
 */
export type Score = Pick<
  UserScore,
  'score' | 'exScore' | 'maxCombo' | 'clearLamp' | 'rank'
>

export function getDanceLevel(score: number) {
  const rankList = [
    { border: 990000, rank: 'AAA' },
    { border: 950000, rank: 'AA+' },
    { border: 900000, rank: 'AA' },
    { border: 890000, rank: 'AA-' },
    { border: 850000, rank: 'A+' },
    { border: 800000, rank: 'A' },
    { border: 790000, rank: 'A-' },
    { border: 750000, rank: 'B+' },
    { border: 700000, rank: 'B' },
    { border: 690000, rank: 'B-' },
    { border: 650000, rank: 'C+' },
    { border: 600000, rank: 'C' },
    { border: 590000, rank: 'C-' },
    { border: 550000, rank: 'D+' },
  ] as const
  for (const { border, rank } of rankList) {
    if (score >= border) return rank
  }
  return 'D'
}
