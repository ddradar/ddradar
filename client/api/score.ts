import type { Api } from '@ddradar/core'
import type { NuxtHTTPInstance } from '@nuxt/http'

import { apiPrefix } from '~/api'

/**
 * Call "Get Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--get/
 */
export function getChartScore(
  $http: Pick<NuxtHTTPInstance, '$get'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4,
  scope?: 'private' | 'medium' | 'full'
) {
  const query = scope ? `?scope=${scope}` : ''
  return $http.$get<Api.ScoreInfo[]>(
    `${apiPrefix}/scores/${songId}/${playStyle}/${difficulty}${query}`
  )
}

/**
 * Call "Post Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--post/
 */
export function postChartScore(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4,
  score: Api.ScoreBody
) {
  return $http.$post(
    `${apiPrefix}/scores/${songId}/${playStyle}/${difficulty}`,
    score
  )
}

/**
 * Call "Delete Chart Score" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id__style__difficulty--delete/
 */
export function deleteChartScore(
  $http: Pick<NuxtHTTPInstance, 'delete'>,
  songId: string,
  playStyle: 1 | 2,
  difficulty: 0 | 1 | 2 | 3 | 4
) {
  return $http.delete(`/api/v1/scores/${songId}/${playStyle}/${difficulty}`)
}

/**
 * Call "Post Song Scores" API.
 * @see https://github.com/ddradar/ddradar/tree/master/api/scores__id--post/
 */
export function postSongScores(
  $http: Pick<NuxtHTTPInstance, '$post'>,
  songId: string,
  scores: Api.ScoreListBody[]
) {
  return $http.$post<Api.ScoreInfo[]>(`${apiPrefix}/scores/${songId}`, scores)
}
