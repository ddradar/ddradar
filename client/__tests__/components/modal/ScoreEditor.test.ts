import { Score } from '@ddradar/core'
import { testSongData } from '@ddradar/core/__tests__/data'
import { createLocalVue, mount, shallowMount, Wrapper } from '@vue/test-utils'
import Buefy from 'buefy'
import { mocked } from 'ts-jest/utils'
import VueI18n from 'vue-i18n'

import { getChartScore, postChartScore } from '~/api/score'
import ScoreEditor from '~/components/modal/ScoreEditor.vue'
import * as popup from '~/utils/popup'

jest.mock('@ddradar/core')
jest.mock('~/api/score')
jest.mock('~/utils/popup')
const localVue = createLocalVue()
localVue.use(Buefy)
localVue.use(VueI18n)

describe('/components/modal/ScoreEditor.vue', () => {
  let wrapper: Wrapper<ScoreEditor>
  const propsData = {
    songId: testSongData.id,
    playStyle: testSongData.charts[0].playStyle,
    difficulty: testSongData.charts[0].difficulty,
    songData: testSongData,
  }
  const i18n = new VueI18n({ locale: 'ja', silentFallbackWarn: true })
  const $parent = { close: jest.fn() }
  beforeEach(() => {
    mocked(getChartScore).mockRejectedValue('404')
    wrapper = shallowMount(ScoreEditor, { localVue, propsData, i18n })
  })

  describe.each(['ja', 'en'])('{ locale: %s } snapshot test', locale => {
    const i18n = new VueI18n({ locale, silentFallbackWarn: true })
    test('{ selectedChart: null } renders chart dropdown', () => {
      const wrapper = mount(ScoreEditor, {
        localVue,
        propsData: { ...propsData, playStyle: null, difficulty: null },
        i18n,
      })
      expect(wrapper).toMatchSnapshot()
    })
    test('{ selectedChart: charts[0] } renders edit fields', () => {
      const wrapper = mount(ScoreEditor, { localVue, propsData, i18n })
      expect(wrapper).toMatchSnapshot()
    })
  })

  // Computed
  describe('get rank()', () => {
    beforeEach(() => mocked(Score.getDanceLevel).mockClear())
    test('{ isFailed: true } returns "E"', () => {
      // Arrange - Act
      wrapper.setData({ isFailed: true })
      // @ts-ignore
      const rank = wrapper.vm.rank

      // Assert
      expect(rank).toBe('E')
      expect(mocked(Score.getDanceLevel)).not.toBeCalled()
    })
    test('{ isFailed: false } returns getDanceLevel()', () => {
      // Arrange
      const score = 800000
      mocked(Score.getDanceLevel).mockClear()
      mocked(Score.getDanceLevel).mockReturnValue('A')

      // Act
      wrapper.setData({ isFailed: false, score })
      // @ts-ignore
      const rank = wrapper.vm.rank

      // Assert
      expect(rank).toBe('A')
      expect(mocked(Score.getDanceLevel)).toBeCalledWith(score)
    })
  })

  // Method
  describe('onChartSelected', () => {
    test('({ playStyle: 1, difficulty: 4 }) sets selectedChart: null', () => {
      // @ts-ignore
      wrapper.vm.onChartSelected({ playStyle: 1, difficulty: 4 })

      // @ts-ignore
      expect(wrapper.vm.selectedChart).toBeNull()
    })
    test('({ playStyle: 1, difficulty: 1 }) sets selectedChart: chart[1]', () => {
      // @ts-ignore
      wrapper.vm.onChartSelected({ playStyle: 1, difficulty: 1 })

      // @ts-ignore
      expect(wrapper.vm.selectedChart).toBe(testSongData.charts[1])
    })
  })
  describe('calcScore', () => {
    beforeEach(() => {
      mocked(Score.setValidScoreFromChart).mockReset()
      mocked(popup.warning).mockClear()
    })

    test('calls setValidScoreFromChart()', () => {
      // Arrange
      mocked(Score.setValidScoreFromChart).mockReturnValue({
        score: 0,
        exScore: 0,
        clearLamp: 0,
        maxCombo: 0,
        rank: 'D',
      })

      // Act
      // @ts-ignore
      wrapper.vm.calcScore()

      // Assert
      expect(mocked(Score.setValidScoreFromChart)).toBeCalled()
    })
    test('calls popup.warning() if setValidScoreFromChart() throws error', () => {
      // Arrange
      mocked(Score.setValidScoreFromChart).mockImplementation(() => {
        throw new Error('Invalid Score')
      })

      // Act
      // @ts-ignore
      wrapper.vm.calcScore()

      // Assert
      expect(mocked(Score.setValidScoreFromChart)).toBeCalled()
      expect(mocked(popup.warning)).toBeCalled()
    })
  })
  describe('saveScore', () => {
    beforeEach(() => {
      mocked(postChartScore).mockClear()
      mocked(popup.success).mockClear()
      mocked(popup.danger).mockClear()
      // @ts-ignore
      wrapper.vm.$parent = $parent
    })
    test('calls postChartScore() and close', async () => {
      // Arrange - Act
      // @ts-ignore
      await wrapper.vm.saveScore()

      // Assert
      expect(mocked(postChartScore)).toBeCalled()
      expect(mocked(popup.success)).toBeCalled()
      expect($parent.close).toBeCalled()
    })
    test('calls popup.danger() if postChartScore() throws error and close', async () => {
      // Arrange
      mocked(postChartScore).mockRejectedValueOnce('Error')

      // Act
      // @ts-ignore
      await wrapper.vm.saveScore()

      // Assert
      expect(mocked(postChartScore)).toBeCalled()
      expect(mocked(popup.success)).not.toBeCalled()
      expect(mocked(popup.danger)).toBeCalled()
      expect($parent.close).toBeCalled()
    })
  })
})
