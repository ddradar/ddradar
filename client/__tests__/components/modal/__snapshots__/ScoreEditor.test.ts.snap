// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/components/modal/ScoreEditor.vue { locale: en } snapshot test { selectedChart: charts[0] } renders edit fields 1`] = `
<div class="modal-card">
  <header class="modal-card-head">
    <h1 class="modal-card-title">PARANOiA</h1>
  </header>
  <section class="modal-card-body">
    <h2 class="subtitle is-small">
      SINGLE/BEGINNER
    </h2>
    <transition-stub name="fade">
      <div class="loading-overlay is-active is-full-page">
        <div class="loading-background"></div>
        <div class="loading-icon"></div>
      </div>
    </transition-stub>
    <div class="field">
      <!---->
      <div class="field-body">
        <div class="field is-grouped">
          <!---->
          <div class="field"><label class="label">Score</label>
            <div class="control is-clearfix"><input type="number" autocomplete="on" required="required" placeholder="0-1000000" min="0" max="1000000" step="10" class="input">
              <!---->
              <!---->
              <!---->
            </div>
            <!---->
          </div> <label class="b-checkbox checkbox"><input type="checkbox" autocomplete="on" true-value="true" value="false"><span class="check"></span><span class="control-label">Rank E</span></label>
          <!---->
        </div>
      </div>
      <!---->
    </div>
    <div class="field"><label class="label">Clear Lamp</label>
      <div class="control"><span class="select"><select><!----><option value="0">Failed</option> <option value="1">Assisted Clear</option> <option value="2">Clear</option> <option value="3">Life4 Clear</option> <option value="4">Good Full Combo</option> <option value="5">Great Full Combo</option> <option value="6">Perfect Full Combo</option> <option value="7">Marvelous Full Combo</option></select></span>
        <!---->
      </div>
      <!---->
    </div>
    <div class="field"><label class="label">EX SCORE</label>
      <div class="control is-clearfix"><input type="number" autocomplete="on" min="0" max="414" placeholder="0-414" class="input">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
    <div class="field"><label class="label">MAX COMBO</label>
      <div class="control is-clearfix"><input type="number" autocomplete="on" min="0" max="138" placeholder="0-138" class="input">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
    <div class="field">
      <!----><button type="button" class="button is-info"><span class="icon is-small"><i class="mdi mdi-calculator"></i></span><span>
          Auto Calc
        </span>
        <!---->
      </button>
      <!---->
    </div>
  </section>
  <footer class="modal-card-foot"><button type="button" class="button is-success"><span class="icon is-small"><i class="mdi mdi-content-save"></i></span><span>
      Save
    </span>
      <!---->
    </button> <button type="button" class="button is-danger"><span class="icon is-small"><i class="mdi mdi-delete"></i></span><span>
      Delete
    </span>
      <!---->
    </button> <button type="button" class="button is-warning">
      <!----><span>
      Close
    </span>
      <!---->
    </button></footer>
</div>
`;

exports[`/components/modal/ScoreEditor.vue { locale: en } snapshot test { selectedChart: null } renders chart dropdown 1`] = `
<div class="modal-card">
  <header class="modal-card-head">
    <h1 class="modal-card-title">PARANOiA</h1>
  </header>
  <section class="modal-card-body">
    <div class="field"><label class="label">Select Chart</label>
      <div class="control"><span class="select is-empty"><select><!----><option value="[object Object]">
          SINGLE/BEGINNER
        </option><option value="[object Object]">
          SINGLE/BASIC
        </option></select></span>
        <!---->
      </div>
      <!---->
    </div>
    <!---->
  </section>
  <!---->
</div>
`;

exports[`/components/modal/ScoreEditor.vue { locale: ja } snapshot test { selectedChart: charts[0] } renders edit fields 1`] = `
<div class="modal-card">
  <header class="modal-card-head">
    <h1 class="modal-card-title">PARANOiA</h1>
  </header>
  <section class="modal-card-body">
    <h2 class="subtitle is-small">
      SINGLE/BEGINNER
    </h2>
    <transition-stub name="fade">
      <div class="loading-overlay is-active is-full-page">
        <div class="loading-background"></div>
        <div class="loading-icon"></div>
      </div>
    </transition-stub>
    <div class="field">
      <!---->
      <div class="field-body">
        <div class="field is-grouped">
          <!---->
          <div class="field"><label class="label">ハイスコア</label>
            <div class="control is-clearfix"><input type="number" autocomplete="on" required="required" placeholder="0-1000000" min="0" max="1000000" step="10" class="input">
              <!---->
              <!---->
              <!---->
            </div>
            <!---->
          </div> <label class="b-checkbox checkbox"><input type="checkbox" autocomplete="on" true-value="true" value="false"><span class="check"></span><span class="control-label">E判定</span></label>
          <!---->
        </div>
      </div>
      <!---->
    </div>
    <div class="field"><label class="label">クリア種別</label>
      <div class="control"><span class="select"><select><!----><option value="0">Failed</option> <option value="1">Assisted Clear</option> <option value="2">Clear</option> <option value="3">Life4 Clear</option> <option value="4">Good Full Combo</option> <option value="5">Great Full Combo</option> <option value="6">Perfect Full Combo</option> <option value="7">Marvelous Full Combo</option></select></span>
        <!---->
      </div>
      <!---->
    </div>
    <div class="field"><label class="label">EXスコア</label>
      <div class="control is-clearfix"><input type="number" autocomplete="on" min="0" max="414" placeholder="0-414" class="input">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
    <div class="field"><label class="label">最大コンボ数</label>
      <div class="control is-clearfix"><input type="number" autocomplete="on" min="0" max="138" placeholder="0-138" class="input">
        <!---->
        <!---->
        <!---->
      </div>
      <!---->
    </div>
    <div class="field">
      <!----><button type="button" class="button is-info"><span class="icon is-small"><i class="mdi mdi-calculator"></i></span><span>
          自動計算
        </span>
        <!---->
      </button>
      <!---->
    </div>
  </section>
  <footer class="modal-card-foot"><button type="button" class="button is-success"><span class="icon is-small"><i class="mdi mdi-content-save"></i></span><span>
      保存
    </span>
      <!---->
    </button> <button type="button" class="button is-danger"><span class="icon is-small"><i class="mdi mdi-delete"></i></span><span>
      削除
    </span>
      <!---->
    </button> <button type="button" class="button is-warning">
      <!----><span>
      閉じる
    </span>
      <!---->
    </button></footer>
</div>
`;

exports[`/components/modal/ScoreEditor.vue { locale: ja } snapshot test { selectedChart: null } renders chart dropdown 1`] = `
<div class="modal-card">
  <header class="modal-card-head">
    <h1 class="modal-card-title">PARANOiA</h1>
  </header>
  <section class="modal-card-body">
    <div class="field"><label class="label">譜面を選択</label>
      <div class="control"><span class="select is-empty"><select><!----><option value="[object Object]">
          SINGLE/BEGINNER
        </option><option value="[object Object]">
          SINGLE/BASIC
        </option></select></span>
        <!---->
      </div>
      <!---->
    </div>
    <!---->
  </section>
  <!---->
</div>
`;
